File: ios-native/Jottery/Jottery/Services/SearchService.swift
import Foundation

/// In-memory search over decrypted notes.
///
/// Supports the full Jottery query syntax for parity with the web and TUI clients:
///
/// - **Tags**: `#tagname` (AND logic), `#tag1 | #tag2` (OR logic)
/// - **Text**: plain words (AND logic), `"exact phrase"`, `word1 | word2` (OR)
/// - **Exclusion**: `-term`, `-#tag`
/// - **Wildcards**: `dog*`, `*dog`, `*dog*`
/// - **Modifiers**: `has:attachment`, `created:>YYYY-MM-DD`, `modified:<YYYY-MM-DD`,
///   `created:YYYY-MM-DD..YYYY-MM-DD`, `words:>N`, `words:<N`, `words:N..N`,
///   `color:name`, `colour:name`, `category:name`
enum SearchService {

    // MARK: - Public API

    /// Filter notes by a search query string.
    static func filter(notes: [DecryptedNote], query: String) -> [DecryptedNote] {
        let trimmed = query.trimmingCharacters(in: .whitespaces)
        guard !trimmed.isEmpty else { return notes }

        let parsed = parseQuery(trimmed)
        return notes.filter { matchesQuery(note: $0, query: parsed) }
    }

    /// Sort notes by the given order, pinned notes always first.
    static func sort(notes: [DecryptedNote], order: SortOrder) -> [DecryptedNote] {
        notes.sorted { a, b in
            if a.pinned != b.pinned { return a.pinned }
            switch order {
            case .recent:  return a.modifiedAt > b.modifiedAt
            case .oldest:  return a.modifiedAt < b.modifiedAt
            case .alpha:   return a.title.localizedCaseInsensitiveCompare(b.title) == .orderedAscending
            case .created: return a.createdAt > b.createdAt
            }
        }
    }

    // MARK: - Parsed Query

    struct ParsedQuery {
        // Tags (AND — all must match)
        var requiredTags: [String] = []
        // Tags (OR groups — at least one in each group must match)
        var orTagGroups: [[String]] = []
        // Tags to exclude
        var excludeTags: [String] = []

        // Text terms (AND — all must match)
        var requiredText: [TextMatch] = []
        // Text terms (OR groups)
        var orTextGroups: [[TextMatch]] = []
        // Text to exclude
        var excludeText: [String] = []

        // Modifiers
        var hasAttachment: Bool?
        var createdAfter: Date?
        var createdBefore: Date?
        var modifiedAfter: Date?
        var modifiedBefore: Date?
        var wordCountMin: Int?
        var wordCountMax: Int?
        var colors: [String] = []
        var excludeColors: [String] = []
    }

    enum TextMatch {
        case contains(String)    // plain word
        case prefix(String)      // word*
        case suffix(String)      // *word
        case infix(String)       // *word*  (same as contains, but from wildcard syntax)
        case exact(String)       // "exact phrase"

        func matches(_ text: String) -> Bool {
            switch self {
            case .contains(let term):  return text.contains(term)
            case .prefix(let term):    return matchesWordPrefix(text, prefix: term)
            case .suffix(let term):    return matchesWordSuffix(text, suffix: term)
            case .infix(let term):     return text.contains(term)
            case .exact(let phrase):   return text.contains(phrase)
            }
        }

        private func matchesWordPrefix(_ text: String, prefix: String) -> Bool {
            // Check if any word in the text starts with the prefix
            let words = text.split(whereSeparator: { $0.isWhitespace || $0.isNewline || $0.isPunctuation })
            return words.contains { String($0).hasPrefix(prefix) }
        }

        private func matchesWordSuffix(_ text: String, suffix: String) -> Bool {
            let words = text.split(whereSeparator: { $0.isWhitespace || $0.isNewline || $0.isPunctuation })
            return words.contains { String($0).hasSuffix(suffix) }
        }
    }

    // MARK: - Query Parsing

    static func parseQuery(_ raw: String) -> ParsedQuery {
        var query = ParsedQuery()
        var remaining = raw

        // 1. Extract modifiers (they get removed from the query string)
        remaining = extractModifiers(from: remaining, into: &query)

        // 2. Tokenize the remainder (handles quotes, #tags, -exclusions, | OR)
        tokenize(remaining, into: &query)

        return query
    }

    // MARK: - Modifier Extraction

    private static let datePattern = #"(\d{4}-\d{2}-\d{2})"#

    private static func extractModifiers(from input: String, into query: inout ParsedQuery) -> String {
        var text = input

        // has:attachment
        if let range = text.range(of: #"\bhas:attachment\b"#, options: .regularExpression) {
            query.hasAttachment = true
            text.removeSubrange(range)
        }

        // created:>DATE
        text = extractDateModifier(from: text, pattern: #"created:>"# + datePattern, setter: { query.createdAfter = $0 })
        // created:<DATE
        text = extractDateModifier(from: text, pattern: #"created:<"# + datePattern, setter: { query.createdBefore = $0 })
        // created:DATE..DATE
        text = extractDateRange(from: text, prefix: "created", setAfter: { query.createdAfter = $0 }, setBefore: { query.createdBefore = $0 })

        // modified:>DATE
        text = extractDateModifier(from: text, pattern: #"modified:>"# + datePattern, setter: { query.modifiedAfter = $0 })
        // modified:<DATE
        text = extractDateModifier(from: text, pattern: #"modified:<"# + datePattern, setter: { query.modifiedBefore = $0 })
        // modified:DATE..DATE
        text = extractDateRange(from: text, prefix: "modified", setAfter: { query.modifiedAfter = $0 }, setBefore: { query.modifiedBefore = $0 })

        // words:>N, words:<N, words:N..N
        text = extractWordCount(from: text, into: &query)

        // color: / colour: / category: filters
        text = extractColorFilters(from: text, into: &query)

        return text.trimmingCharacters(in: .whitespaces)
    }

    private static func extractDateModifier(from text: String, pattern: String, setter: (Date) -> Void) -> String {
        var result = text
        if let match = result.range(of: pattern, options: .regularExpression) {
            let matched = String(result[match])
            // Extract the date part (last 10 chars = YYYY-MM-DD)
            let dateStr = String(matched.suffix(10))
            if let date = parseDate(dateStr) {
                setter(date)
            }
            result.removeSubrange(match)
        }
        return result
    }

    private static func extractDateRange(from text: String, prefix: String, setAfter: (Date) -> Void, setBefore: (Date) -> Void) -> String {
        var result = text
        let pattern = prefix + #":"# + datePattern + #"\.\."# + datePattern
        if let match = result.range(of: pattern, options: .regularExpression) {
            let matched = String(result[match])
            // Extract both dates
            let dateRegex = try? NSRegularExpression(pattern: datePattern)
            let nsRange = NSRange(matched.startIndex..<matched.endIndex, in: matched)
            let matches = dateRegex?.matches(in: matched, range: nsRange) ?? []
            if matches.count >= 2 {
                if let r1 = Range(matches[0].range, in: matched), let d1 = parseDate(String(matched[r1])) {
                    setAfter(d1)
                }
                if let r2 = Range(matches[1].range, in: matched), let d2 = parseDate(String(matched[r2])) {
                    // Set before to end of day
                    setBefore(Calendar.current.date(byAdding: .day, value: 1, to: d2) ?? d2)
                }
            }
            result.removeSubrange(match)
        }
        return result
    }

    private static func extractWordCount(from text: String, into query: inout ParsedQuery) -> String {
        var result = text

        // words:>N
        if let match = result.range(of: #"words:>(\d+)"#, options: .regularExpression) {
            let matched = String(result[match])
            if let num = Int(matched.replacingOccurrences(of: "words:>", with: "")) {
                query.wordCountMin = num
            }
            result.removeSubrange(match)
        }

        // words:<N
        if let match = result.range(of: #"words:<(\d+)"#, options: .regularExpression) {
            let matched = String(result[match])
            if let num = Int(matched.replacingOccurrences(of: "words:<", with: "")) {
                query.wordCountMax = num
            }
            result.removeSubrange(match)
        }

        // words:N..N
        if let match = result.range(of: #"words:(\d+)\.\.(\d+)"#, options: .regularExpression) {
            let matched = String(result[match])
            let nums = matched.replacingOccurrences(of: "words:", with: "").split(separator: ".").compactMap { Int($0) }
            if nums.count >= 2 {
                query.wordCountMin = nums[0]
                query.wordCountMax = nums[1]
            }
            result.removeSubrange(match)
        }

        return result
    }

    private static func extractColorFilters(from text: String, into query: inout ParsedQuery) -> String {
        var result = text

        // Match color:name, colour:name, or category:name (case-insensitive)
        let colorPattern = #"(?i)\b(?:colou?r|category):([a-z0-9#]+)\b"#
        while let match = result.range(of: colorPattern, options: .regularExpression) {
            let matched = String(result[match])
            // Extract the color name after the colon
            if let colonIdx = matched.firstIndex(of: ":") {
                let colorName = String(matched[matched.index(after: colonIdx)...]).lowercased()
                query.colors.append(colorName)
            }
            result.removeSubrange(match)
        }

        return result
    }

    // MARK: - Tokenization

    private static func tokenize(_ input: String, into query: inout ParsedQuery) {
        let tokens = splitTokens(input)

        // Group tokens by | (OR) separators
        var currentGroup: [RawToken] = []
        var groups: [[RawToken]] = []
        var hasOr = false

        for token in tokens {
            if case .orSeparator = token {
                hasOr = true
                groups.append(currentGroup)
                currentGroup = []
            } else {
                currentGroup.append(token)
            }
        }
        groups.append(currentGroup)

        if hasOr && groups.count > 1 {
            // Check if this is an OR group of tags or text
            let allTags = groups.allSatisfy { group in
                group.count == 1 && {
                    if case .tag = group[0] { return true }
                    return false
                }()
            }

            if allTags {
                // OR tag group
                let tagNames = groups.compactMap { group -> String? in
                    guard let token = group.first, case .tag(let name) = token else { return nil }
                    return name
                }
                query.orTagGroups.append(tagNames)
            } else {
                // OR text group
                let textMatches = groups.compactMap { group -> TextMatch? in
                    guard let token = group.first else { return nil }
                    return textMatchFromToken(token)
                }
                if !textMatches.isEmpty {
                    query.orTextGroups.append(textMatches)
                }
            }
        } else {
            // No OR — treat as AND
            for token in currentGroup {
                switch token {
                case .tag(let name):
                    query.requiredTags.append(name)
                case .excludeTag(let name):
                    query.excludeTags.append(name)
                case .exclude(let term):
                    query.excludeText.append(term)
                case .text(let term):
                    query.requiredText.append(classifyTextMatch(term))
                case .phrase(let phrase):
                    query.requiredText.append(.exact(phrase.lowercased()))
                case .orSeparator:
                    break
                }
            }
        }
    }

    private enum RawToken {
        case tag(String)
        case excludeTag(String)
        case exclude(String)
        case text(String)
        case phrase(String)
        case orSeparator
    }

    private static func splitTokens(_ input: String) -> [RawToken] {
        var tokens: [RawToken] = []
        var i = input.startIndex

        while i < input.endIndex {
            // Skip whitespace
            if input[i].isWhitespace {
                i = input.index(after: i)
                continue
            }

            // OR separator
            if input[i] == "|" {
                tokens.append(.orSeparator)
                i = input.index(after: i)
                continue
            }

            // Quoted phrase
            if input[i] == "\"" {
                let start = input.index(after: i)
                if let end = input[start...].firstIndex(of: "\"") {
                    let phrase = String(input[start..<end])
                    tokens.append(.phrase(phrase))
                    i = input.index(after: end)
                } else {
                    // No closing quote — treat rest as phrase
                    let phrase = String(input[start...])
                    tokens.append(.phrase(phrase))
                    i = input.endIndex
                }
                continue
            }

            // Read a word (until whitespace or |)
            let wordStart = i
            while i < input.endIndex && !input[i].isWhitespace && input[i] != "|" {
                i = input.index(after: i)
            }
            let word = String(input[wordStart..<i])

            if word.hasPrefix("-#") && word.count > 2 {
                tokens.append(.excludeTag(String(word.dropFirst(2)).lowercased()))
            } else if word.hasPrefix("#") && word.count > 1 {
                tokens.append(.tag(String(word.dropFirst()).lowercased()))
            } else if word.hasPrefix("-") && word.count > 1 {
                tokens.append(.exclude(String(word.dropFirst()).lowercased()))
            } else if !word.isEmpty {
                tokens.append(.text(word.lowercased()))
            }
        }

        return tokens
    }

    private static func classifyTextMatch(_ term: String) -> TextMatch {
        if term.hasPrefix("*") && term.hasSuffix("*") && term.count > 2 {
            return .infix(String(term.dropFirst().dropLast()))
        } else if term.hasSuffix("*") && term.count > 1 {
            return .prefix(String(term.dropLast()))
        } else if term.hasPrefix("*") && term.count > 1 {
            return .suffix(String(term.dropFirst()))
        }
        return .contains(term)
    }

    private static func textMatchFromToken(_ token: RawToken) -> TextMatch? {
        switch token {
        case .text(let term): return classifyTextMatch(term)
        case .phrase(let phrase): return .exact(phrase.lowercased())
        case .tag: return nil
        case .excludeTag, .exclude: return nil
        case .orSeparator: return nil
        }
    }

    // MARK: - Matching

    private static func matchesQuery(note: DecryptedNote, query: ParsedQuery) -> Bool {
        let lowerContent = note.content.lowercased()
        let lowerTags = note.tags.map { $0.lowercased() }
        let searchableText = lowerContent + " " + lowerTags.joined(separator: " ")

        // Required tags (AND)
        for tag in query.requiredTags {
            if !lowerTags.contains(tag) { return false }
        }

        // OR tag groups
        for group in query.orTagGroups {
            if !group.contains(where: { lowerTags.contains($0) }) { return false }
        }

        // Excluded tags
        for tag in query.excludeTags {
            if lowerTags.contains(tag) { return false }
        }

        // Required text (AND)
        for match in query.requiredText {
            if !match.matches(searchableText) { return false }
        }

        // OR text groups
        for group in query.orTextGroups {
            if !group.contains(where: { $0.matches(searchableText) }) { return false }
        }

        // Excluded text
        for term in query.excludeText {
            if lowerContent.contains(term) { return false }
        }

        // has:attachment
        if let hasAttachment = query.hasAttachment, hasAttachment {
            if note.attachments.isEmpty { return false }
        }

        // Date filters
        if let after = query.createdAfter, note.createdAt < after { return false }
        if let before = query.createdBefore, note.createdAt > before { return false }
        if let after = query.modifiedAfter, note.modifiedAt < after { return false }
        if let before = query.modifiedBefore, note.modifiedAt > before { return false }

        // Word count
        if let min = query.wordCountMin, note.wordCount < min { return false }
        if let max = query.wordCountMax, note.wordCount > max { return false }

        // Color / category
        if !query.colors.isEmpty {
            guard let noteColor = note.color?.lowercased() else { return false }
            if !query.colors.contains(noteColor) { return false }
        }

        return true
    }

    // MARK: - Date Parsing

    private static let dateFormatter: DateFormatter = {
        let fmt = DateFormatter()
        fmt.dateFormat = "yyyy-MM-dd"
        fmt.locale = Locale(identifier: "en_US_POSIX")
        fmt.timeZone = .current
        return fmt
    }()

    private static func parseDate(_ string: String) -> Date? {
        dateFormatter.date(from: string)
    }
}

File: android-native/app/src/main/java/com/jottery/android/service/SearchService.kt
package com.jottery.android.service

import com.jottery.android.model.DecryptedNote
import com.jottery.android.model.SortOrder
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.time.format.DateTimeParseException

/**
 * In-memory search over decrypted notes.
 *
 * Supports the full Jottery query syntax for parity with the web, TUI, and iOS clients:
 *
 * - **Tags**: `#tagname` (AND logic), `#tag1 | #tag2` (OR logic)
 * - **Exclusion**: `-term`, `-#tag`
 * - **Text**: plain words (AND logic), `"exact phrase"`, `word1 | word2` (OR)
 * - **Wildcards**: `dog*`, `*dog`, `*dog*`
 * - **Modifiers**: `has:attachment`, `created:>YYYY-MM-DD`, `modified:<YYYY-MM-DD`,
 *   `created:YYYY-MM-DD..YYYY-MM-DD`, `words:>N`, `words:<N`, `words:N..N`,
 *   `color:name`, `colour:name`, `category:name`
 */
object SearchService {

    // ------------------------------------------------------------------ //
    //  Public API
    // ------------------------------------------------------------------ //

    /**
     * Result of a filter operation, including the match count for UI display.
     */
    data class FilterResult(
        val notes: List<DecryptedNote>,
        val matchCount: Int,
        val totalCount: Int,
    )

    /**
     * Filter notes by a search query string.
     *
     * Returns a [FilterResult] containing the matched notes plus counts
     * for the search bar display (e.g. "87/456").
     */
    fun filter(notes: List<DecryptedNote>, query: String): FilterResult {
        val trimmed = query.trim()
        if (trimmed.isEmpty()) {
            return FilterResult(notes = notes, matchCount = notes.size, totalCount = notes.size)
        }

        val parsed = parseQuery(trimmed)
        val filtered = notes.filter { matchesQuery(it, parsed) }
        return FilterResult(notes = filtered, matchCount = filtered.size, totalCount = notes.size)
    }

    /**
     * Sort notes by the given order, with pinned notes always first.
     */
    fun sort(notes: List<DecryptedNote>, order: SortOrder): List<DecryptedNote> {
        return notes.sortedWith(Comparator { a, b ->
            // Pinned notes always come first
            if (a.pinned != b.pinned) {
                return@Comparator if (a.pinned) -1 else 1
            }
            when (order) {
                SortOrder.RECENT -> b.modifiedAt.compareTo(a.modifiedAt)
                SortOrder.OLDEST -> a.modifiedAt.compareTo(b.modifiedAt)
                SortOrder.ALPHA -> a.title.compareTo(b.title, ignoreCase = true)
                SortOrder.CREATED -> b.createdAt.compareTo(a.createdAt)
            }
        })
    }

    // ------------------------------------------------------------------ //
    //  Parsed Query
    // ------------------------------------------------------------------ //

    data class ParsedQuery(
        // Tags (AND -- all must match)
        val requiredTags: MutableList<String> = mutableListOf(),
        // Tags (OR groups -- at least one in each group must match)
        val orTagGroups: MutableList<List<String>> = mutableListOf(),
        // Tags to exclude
        val excludeTags: MutableList<String> = mutableListOf(),

        // Text terms (AND -- all must match)
        val requiredText: MutableList<TextMatch> = mutableListOf(),
        // Text terms (OR groups)
        val orTextGroups: MutableList<List<TextMatch>> = mutableListOf(),
        // Text to exclude
        val excludeText: MutableList<String> = mutableListOf(),

        // Modifiers
        var hasAttachment: Boolean? = null,
        var createdAfter: Instant? = null,
        var createdBefore: Instant? = null,
        var modifiedAfter: Instant? = null,
        var modifiedBefore: Instant? = null,
        var wordCountMin: Int? = null,
        var wordCountMax: Int? = null,
        val colors: MutableList<String> = mutableListOf(),
        val excludeColors: MutableList<String> = mutableListOf(),
    )

    sealed class TextMatch {
        /** Plain word -- case-insensitive contains. */
        data class Contains(val term: String) : TextMatch()
        /** word* -- matches words starting with prefix. */
        data class Prefix(val term: String) : TextMatch()
        /** *word -- matches words ending with suffix. */
        data class Suffix(val term: String) : TextMatch()
        /** *word* -- same as contains, but from explicit wildcard syntax. */
        data class Infix(val term: String) : TextMatch()
        /** "exact phrase" -- case-insensitive phrase match. */
        data class Exact(val phrase: String) : TextMatch()

        fun matches(text: String): Boolean = when (this) {
            is Contains -> text.contains(term)
            is Prefix -> matchesWordPrefix(text, term)
            is Suffix -> matchesWordSuffix(text, term)
            is Infix -> text.contains(term)
            is Exact -> text.contains(phrase)
        }

        private fun matchesWordPrefix(text: String, prefix: String): Boolean {
            // Check if any word in the text starts with the prefix
            return text.split(WORD_SPLIT_REGEX).any { word ->
                word.isNotEmpty() && word.startsWith(prefix)
            }
        }

        private fun matchesWordSuffix(text: String, suffix: String): Boolean {
            return text.split(WORD_SPLIT_REGEX).any { word ->
                word.isNotEmpty() && word.endsWith(suffix)
            }
        }

        companion object {
            private val WORD_SPLIT_REGEX = Regex("[\\s\\n\\p{Punct}]+")
        }
    }

    // ------------------------------------------------------------------ //
    //  Query Parsing
    // ------------------------------------------------------------------ //

    fun parseQuery(raw: String): ParsedQuery {
        val query = ParsedQuery()
        // 1. Extract modifiers (removed from the query string)
        val remaining = extractModifiers(raw, query)
        // 2. Tokenize the remainder (handles quotes, #tags, -exclusions, | OR)
        tokenize(remaining, query)
        return query
    }

    // ------------------------------------------------------------------ //
    //  Modifier Extraction
    // ------------------------------------------------------------------ //

    private val DATE_PATTERN = """(\d{4}-\d{2}-\d{2})"""
    private val DATE_REGEX = Regex(DATE_PATTERN)

    private fun extractModifiers(input: String, query: ParsedQuery): String {
        var text = input

        // has:attachment
        val hasAttachmentRegex = Regex("""\bhas:attachment\b""")
        if (hasAttachmentRegex.containsMatchIn(text)) {
            query.hasAttachment = true
            text = hasAttachmentRegex.replaceFirst(text, "")
        }

        // created:>DATE
        text = extractDateModifier(text, Regex("""created:>$DATE_PATTERN""")) { query.createdAfter = it }
        // created:<DATE
        text = extractDateModifier(text, Regex("""created:<$DATE_PATTERN""")) { query.createdBefore = it }
        // created:DATE..DATE
        text = extractDateRange(text, "created", { query.createdAfter = it }, { query.createdBefore = it })

        // modified:>DATE
        text = extractDateModifier(text, Regex("""modified:>$DATE_PATTERN""")) { query.modifiedAfter = it }
        // modified:<DATE
        text = extractDateModifier(text, Regex("""modified:<$DATE_PATTERN""")) { query.modifiedBefore = it }
        // modified:DATE..DATE
        text = extractDateRange(text, "modified", { query.modifiedAfter = it }, { query.modifiedBefore = it })

        // words:>N, words:<N, words:N..N
        text = extractWordCount(text, query)

        // color: / colour: / category: filters
        text = extractColorFilters(text, query)

        return text.trim()
    }

    private fun extractDateModifier(
        text: String,
        pattern: Regex,
        setter: (Instant) -> Unit,
    ): String {
        val match = pattern.find(text) ?: return text
        val matched = match.value
        // The date is always the last 10 characters (YYYY-MM-DD)
        val dateStr = matched.takeLast(10)
        val date = parseDate(dateStr)
        if (date != null) {
            setter(date)
        }
        return text.removeRange(match.range)
    }

    private fun extractDateRange(
        text: String,
        prefix: String,
        setAfter: (Instant) -> Unit,
        setBefore: (Instant) -> Unit,
    ): String {
        val pattern = Regex("""$prefix:$DATE_PATTERN\.\.$DATE_PATTERN""")
        val match = pattern.find(text) ?: return text
        val matched = match.value

        val dates = DATE_REGEX.findAll(matched).toList()
        if (dates.size >= 2) {
            val d1 = parseDate(dates[0].value)
            val d2 = parseDate(dates[1].value)
            if (d1 != null) setAfter(d1)
            if (d2 != null) {
                // Set before to end of day (start of next day)
                val nextDay = LocalDate.parse(dates[1].value, DATE_FORMATTER).plusDays(1)
                setBefore(nextDay.atStartOfDay(ZoneId.systemDefault()).toInstant())
            }
        }
        return text.removeRange(match.range)
    }

    private fun extractWordCount(text: String, query: ParsedQuery): String {
        var result = text

        // words:>N
        val gtPattern = Regex("""words:>(\d+)""")
        gtPattern.find(result)?.let { match ->
            val num = match.groupValues[1].toIntOrNull()
            if (num != null) query.wordCountMin = num
            result = result.removeRange(match.range)
        }

        // words:<N
        val ltPattern = Regex("""words:<(\d+)""")
        ltPattern.find(result)?.let { match ->
            val num = match.groupValues[1].toIntOrNull()
            if (num != null) query.wordCountMax = num
            result = result.removeRange(match.range)
        }

        // words:N..N
        val rangePattern = Regex("""words:(\d+)\.\.(\d+)""")
        rangePattern.find(result)?.let { match ->
            val min = match.groupValues[1].toIntOrNull()
            val max = match.groupValues[2].toIntOrNull()
            if (min != null) query.wordCountMin = min
            if (max != null) query.wordCountMax = max
            result = result.removeRange(match.range)
        }

        return result
    }

    private fun extractColorFilters(text: String, query: ParsedQuery): String {
        var result = text

        // Match color:name, colour:name, or category:name (case-insensitive)
        val colorPattern = Regex("""(?i)\b(?:colou?r|category):([a-z0-9#]+)\b""")
        var match = colorPattern.find(result)
        while (match != null) {
            val colorName = match.groupValues[1].lowercase()
            query.colors.add(colorName)
            result = result.removeRange(match.range)
            match = colorPattern.find(result)
        }

        return result
    }

    // ------------------------------------------------------------------ //
    //  Tokenization
    // ------------------------------------------------------------------ //

    private sealed class RawToken {
        data class Tag(val name: String) : RawToken()
        data class ExcludeTag(val name: String) : RawToken()
        data class Exclude(val term: String) : RawToken()
        data class Text(val term: String) : RawToken()
        data class Phrase(val phrase: String) : RawToken()
        data object OrSeparator : RawToken()
    }

    private fun tokenize(input: String, query: ParsedQuery) {
        val tokens = splitTokens(input)

        // Group tokens by | (OR) separators
        var currentGroup = mutableListOf<RawToken>()
        val groups = mutableListOf<List<RawToken>>()
        var hasOr = false

        for (token in tokens) {
            if (token is RawToken.OrSeparator) {
                hasOr = true
                groups.add(currentGroup)
                currentGroup = mutableListOf()
            } else {
                currentGroup.add(token)
            }
        }
        groups.add(currentGroup)

        if (hasOr && groups.size > 1) {
            // Check if this is an OR group of tags or text
            val allTags = groups.all { group ->
                group.size == 1 && group[0] is RawToken.Tag
            }

            if (allTags) {
                // OR tag group
                val tagNames = groups.mapNotNull { group ->
                    val token = group.firstOrNull()
                    if (token is RawToken.Tag) token.name else null
                }
                query.orTagGroups.add(tagNames)
            } else {
                // OR text group
                val textMatches = groups.mapNotNull { group ->
                    val token = group.firstOrNull() ?: return@mapNotNull null
                    textMatchFromToken(token)
                }
                if (textMatches.isNotEmpty()) {
                    query.orTextGroups.add(textMatches)
                }
            }
        } else {
            // No OR -- treat as AND
            for (token in currentGroup) {
                when (token) {
                    is RawToken.Tag -> query.requiredTags.add(token.name)
                    is RawToken.ExcludeTag -> query.excludeTags.add(token.name)
                    is RawToken.Exclude -> query.excludeText.add(token.term)
                    is RawToken.Text -> query.requiredText.add(classifyTextMatch(token.term))
                    is RawToken.Phrase -> query.requiredText.add(TextMatch.Exact(token.phrase.lowercase()))
                    is RawToken.OrSeparator -> { /* already handled */ }
                }
            }
        }
    }

    private fun splitTokens(input: String): List<RawToken> {
        val tokens = mutableListOf<RawToken>()
        var i = 0

        while (i < input.length) {
            val ch = input[i]

            // Skip whitespace
            if (ch.isWhitespace()) {
                i++
                continue
            }

            // OR separator
            if (ch == '|') {
                tokens.add(RawToken.OrSeparator)
                i++
                continue
            }

            // Quoted phrase
            if (ch == '"') {
                val start = i + 1
                val end = input.indexOf('"', start)
                if (end != -1) {
                    val phrase = input.substring(start, end)
                    tokens.add(RawToken.Phrase(phrase))
                    i = end + 1
                } else {
                    // No closing quote -- treat rest as phrase
                    val phrase = input.substring(start)
                    tokens.add(RawToken.Phrase(phrase))
                    i = input.length
                }
                continue
            }

            // Read a word (until whitespace or |)
            val wordStart = i
            while (i < input.length && !input[i].isWhitespace() && input[i] != '|') {
                i++
            }
            val word = input.substring(wordStart, i)

            when {
                word.startsWith("-#") && word.length > 2 ->
                    tokens.add(RawToken.ExcludeTag(word.drop(2).lowercase()))
                word.startsWith("#") && word.length > 1 ->
                    tokens.add(RawToken.Tag(word.drop(1).lowercase()))
                word.startsWith("-") && word.length > 1 ->
                    tokens.add(RawToken.Exclude(word.drop(1).lowercase()))
                word.isNotEmpty() ->
                    tokens.add(RawToken.Text(word.lowercase()))
            }
        }

        return tokens
    }

    private fun classifyTextMatch(term: String): TextMatch {
        return when {
            term.startsWith("*") && term.endsWith("*") && term.length > 2 ->
                TextMatch.Infix(term.drop(1).dropLast(1))
            term.endsWith("*") && term.length > 1 ->
                TextMatch.Prefix(term.dropLast(1))
            term.startsWith("*") && term.length > 1 ->
                TextMatch.Suffix(term.drop(1))
            else ->
                TextMatch.Contains(term)
        }
    }

    private fun textMatchFromToken(token: RawToken): TextMatch? = when (token) {
        is RawToken.Text -> classifyTextMatch(token.term)
        is RawToken.Phrase -> TextMatch.Exact(token.phrase.lowercase())
        is RawToken.Tag -> null
        is RawToken.ExcludeTag -> null
        is RawToken.Exclude -> null
        is RawToken.OrSeparator -> null
    }

    // ------------------------------------------------------------------ //
    //  Matching
    // ------------------------------------------------------------------ //

    private fun matchesQuery(note: DecryptedNote, query: ParsedQuery): Boolean {
        val lowerContent = note.content.lowercase()
        val lowerTags = note.tags.map { it.lowercase() }
        val searchableText = lowerContent + " " + lowerTags.joinToString(" ")

        // Required tags (AND)
        for (tag in query.requiredTags) {
            if (tag !in lowerTags) return false
        }

        // OR tag groups
        for (group in query.orTagGroups) {
            if (group.none { it in lowerTags }) return false
        }

        // Excluded tags
        for (tag in query.excludeTags) {
            if (tag in lowerTags) return false
        }

        // Required text (AND)
        for (match in query.requiredText) {
            if (!match.matches(searchableText)) return false
        }

        // OR text groups
        for (group in query.orTextGroups) {
            if (group.none { it.matches(searchableText) }) return false
        }

        // Excluded text
        for (term in query.excludeText) {
            if (lowerContent.contains(term)) return false
        }

        // has:attachment
        if (query.hasAttachment == true) {
            if (note.attachments.isEmpty()) return false
        }

        // Date filters
        if (query.createdAfter != null && note.createdAt.isBefore(query.createdAfter)) return false
        if (query.createdBefore != null && note.createdAt.isAfter(query.createdBefore)) return false
        if (query.modifiedAfter != null && note.modifiedAt.isBefore(query.modifiedAfter)) return false
        if (query.modifiedBefore != null && note.modifiedAt.isAfter(query.modifiedBefore)) return false

        // Word count
        if (query.wordCountMin != null && note.wordCount < query.wordCountMin!!) return false
        if (query.wordCountMax != null && note.wordCount > query.wordCountMax!!) return false

        // Color / category
        if (query.colors.isNotEmpty()) {
            val noteColor = note.color?.lowercase() ?: return false
            if (noteColor !in query.colors) return false
        }

        return true
    }

    // ------------------------------------------------------------------ //
    //  Date Parsing
    // ------------------------------------------------------------------ //

    private val DATE_FORMATTER: DateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd")

    /**
     * Parse a date string (YYYY-MM-DD) into an [Instant] at the start of that day
     * in the system default time zone.
     */
    private fun parseDate(dateStr: String): Instant? {
        return try {
            val localDate = LocalDate.parse(dateStr, DATE_FORMATTER)
            localDate.atStartOfDay(ZoneId.systemDefault()).toInstant()
        } catch (_: DateTimeParseException) {
            null
        }
    }
}
